librad/src/id/entity.rs

```diff
@@ -0,0 +1,515 @@
+
+pub mod data;
+
+use crate::{
+    id::{uri::RadicleUri, user::User},
+    keys::device::{Key, PublicKey, Signature},
+};
+use async_trait::async_trait;
+use futures::stream::{Stream, StreamExt};
+use multihash::{Multihash, Sha2_256};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    collections::{HashMap, HashSet},
+    iter::FromIterator,
+    str::FromStr,
+};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Serialization failed ({0})")]
+    SerializationFailed(serde_json::error::Error),
+
+    #[error("Invalid UTF8 ({0})")]
+    InvalidUtf8(std::string::FromUtf8Error),
+
+    #[error("Invalid buffer encoding ({0})")]
+    InvalidBufferEncoding(String),
+
+    #[error("Invalid hash ({0})")]
+    InvalidHash(String),
+
+    #[error("Invalid URI ({0})")]
+    InvalidUri(String),
+
+    #[error("Signature already present ({0})")]
+    SignatureAlreadyPresent(PublicKey),
+
+    #[error("Invalid data ({0})")]
+    InvalidData(String),
+
+    #[error("Key not present ({0})")]
+    KeyNotPresent(PublicKey),
+
+    #[error("User not present ({0})")]
+    UserNotPresent(RadicleUri),
+
+    #[error("User key not present (uri {0}, key {1})")]
+    UserKeyNotPresent(RadicleUri, PublicKey),
+
+    #[error("User key not present")]
+    UserKeyNotPresentWIP(RadicleUri, PublicKey),
+
+    #[error("Signature missing")]
+    SignatureMissing,
+
+    #[error("Signature decoding failed")]
+    SignatureDecodingFailed,
+
+    #[error("Signature verification failed")]
+    SignatureVerificationFailed,
+
+    #[error("Resolution failed (uri {0})")]
+    ResolutionFailed(String),
+}
+
+#[derive(Debug, Error)]
+pub enum UpdateVerificationError {
+    #[error("Non monotonic revision")]
+    NonMonotonicRevision,
+
+    #[error("Wrong parent hash")]
+    WrongParentHash,
+
+    #[error("Update without previous quorum")]
+    NoPreviousQuorum,
+
+    #[error("Update without current quorum")]
+    NoCurrentQuorum,
+}
+
+#[derive(Debug, Error)]
+pub enum HistoryVerificationError {
+    #[error("Empty history")]
+    EmptyHistory,
+
+    #[error("Error at revsion (rev {revision:?}, err {error:?})")]
+    ErrorAtRevision { revision: u64, error: Error },
+
+    #[error("Update error (rev {revision:?}, err {error:?})")]
+    UpdateError {
+        revision: u64,
+        error: UpdateVerificationError,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub enum Signatory {
+    User(RadicleUri),
+    OwnedKey,
+}
+
+#[derive(Clone, Debug)]
+pub struct EntitySignature {
+    pub by: Signatory,
+    pub sig: Signature,
+}
+
+#[async_trait]
+pub trait Resolver<T> {
+    async fn resolve(&self, uri: &RadicleUri) -> Result<T, Error>;
+}
+
+// Sized
+#[derive(Clone)]
+pub struct Entity<T> {
+    name: String,
+    revision: u64,
+    hash: Multihash,
+    parent_hash: Option<Multihash>,
+    signatures: HashMap<PublicKey, EntitySignature>,
+    keys: HashSet<PublicKey>,
+    certifiers: HashSet<RadicleUri>,
+    info: T,
+}
+
+impl<T> Entity<T>
```

Comment:
It's a bit unclear what the actual (crate-)public entry points are: the `check_*` methods seem more like private helpers, and could use some docs and/or more descriptive names. I also don't quite understand when we will deserialise the `EntityData` (and check the hash).